### ISSUE TITLE
Having an issue with Harvesters in game causing a LUA Error

### ISSUE THREAD

**Issue by cwattyeso** [Courseplay/courseplay#5038]
I have loaded up Farm Sim today after about a month of not playing. Was using Courseplay Version 362, and had a serious error when using a Harvester. Came to the GitHub and now see Courseplay has been updated to Version 395 since I last played. Downloaded and installed the new version and deleted the old version from my Mod Folder.

Loaded up my save game, created a new course for the field I am wanting to harvest. This course contains two harvesters working together, skipping a lane and only working on up & down rows, no headlands involved. Harvesters are set to lower tool early and raise tool late, and I'm trying to get them to auto unload themselves. I have a truck parked on either side of the field with a trailer.

I start both harvesters off and they set about harvesting perfectly fine, until they reach the point where they hit 100% full. At that point the camera in game stops moving, but the harvester continues and basically drives off screen. I can't tab to other vehicles or get out of the machine. I can stop it with the Courseplay HUD if I have it open on screen.

When I bring up the console in game or check the log after closing out I'm seeing this error:

Removed logmess - Tensuko

**Comment by Tensuko** [comment]
This is not how to post a logfile...
Drag and drop thelog.txt into commentbox, anything else is just horrible to read.

**Comment by cwattyeso** [comment]
I'll do that but will need to trim down my log file as it's rather large. I thought I'd just post the bit relevant to the error and CoursePlay from the end.

**Comment by cwattyeso** [comment]
Here is my log, with everything removed prior to me updating courseplay and playing today and running into the error.
[courseplay error log.txt](https://github.com/Courseplay/courseplay/files/4313681/courseplay.error.log.txt)

**Comment by Tensuko** [comment]
I am not able to reproduce this, need more Information:
- Map
- Field
- Harvester
- Screenshot of Course
- What Settings for Harvester

**Comment by cwattyeso** [comment]
Map is the Fenton Forest 4X Map by Stevie
Field is Field 25
Harvester is the Tribine

Screenshot of the Course:
![fsScreen_2020_03_10_17_58_00](https://user-images.githubusercontent.com/13921716/76343869-9fc32800-62f8-11ea-9c96-a7cabe0d34a9.png)

Harvester Settings:
![fsScreen_2020_03_10_17_58_14](https://user-images.githubusercontent.com/13921716/76343904-a9e52680-62f8-11ea-8804-48ca50eb61bf.png)
![fsScreen_2020_03_10_17_58_20](https://user-images.githubusercontent.com/13921716/76343920-ac478080-62f8-11ea-83e8-dde7ea13a111.png)
![fsScreen_2020_03_10_17_58_30](https://user-images.githubusercontent.com/13921716/76343932-b0739e00-62f8-11ea-81a4-65de27281460.png)

Advanced Settings:
![fsScreen_2020_03_10_17_58_37](https://user-images.githubusercontent.com/13921716/76343945-b36e8e80-62f8-11ea-861e-fab7e117d27f.png)

**Comment by cwattyeso** [comment]
To rule out it being a Mod Conflict Issue, I just tried to harvest Field 1 on the same map using Giants own Claas Lexion Harvester, using single harvester only, up and down rows course generation. Truck parked on the left hand side of the field. All the options selected that I think I need within the Harvester settings and set it off. It Harvested up and down fine for 2 and a Half rows, got full and did the error again. 
[log.txt](https://github.com/Courseplay/courseplay/files/4314050/log.txt)


I'm guessing it just does not like the Auto Unload Option, and I'm just going to have manually unload harvesters. Unless I'm missing some instructions or tutorial on how to get the combines to actual unload themselves.

**Comment by Tensuko** [comment]
![fsScreen_2020_03_10_18_45_19](https://user-images.githubusercontent.com/27344744/76346031-a0fa5200-6304-11ea-9733-4a40630f5479.png)
Pathfind turn while another Harvester goes for self unload.
What Trailer do you have set up for unload ?

**Comment by cwattyeso** [comment]
I have a tried with two different trailers. The Semi-Tipper 70000/6 on the ModHub and the Wilson Pacesetter Super B also on the in game ModHub.

Now one thing I should add, I've not done anything with the trucks, they have no courses loaded on them or anything. I just have driven them and parked them on the edge of the respective field. I have removed the covers before I jumped out, so they should be available for the Harvesters to empty into.

**Comment by Merrlin** [comment]
Just to clarify. The right side of the Trailers are pointing in the direction of the field?

**Comment by cwattyeso** [comment]
Yes the right hand side of the trailers is placed exactly on the field edge, I've even tried testing it this evening with the truck and trailer parked on the field on a strip the harvester has already harvested, so can definitely get to the truck.

Doesn't seem to matter what combination of vehicle/trailer I pick or what Harvester I try to use, as soon as the Harvester reaches 100% Full, that's when it starts throwing that LUA Error and it messes the game up. It won't save progress after that point. If I press the tilde key to open the console and try type in the command to stop all courseplay drivers, I can't do that as no matter how many times I press the tilde key I don't get the part of the console come up that I can enter commands into. That's the first time I've ever had that happen. I've usually always in the past been able to use the console to stop CoursePlay when it starts fouling up and erroring out. This error though seems to break the console, leaving Alt+F4 my only option.

**Comment by cwattyeso** [comment]
Okay now I am confused totally, as just as a test created a new mod folder, downloaded a new map and a couple of mods into, threw in the latest Courseplay I downloaded. Loaded up the game, started a brand new save on this new map. Went straight to the first field needing harvesting, set up a Harvester with Courseplay and parked whatever tractor and trailer the map gave me to start with next to the field. Hit start at first waypoint on the harvester and sat back to observe. It harvested the whole field, emptying itself twice during the course of the field, and then one more when it reached the end of the field, before announcing it had reached it's end point once it had emptied.
[log.txt](https://github.com/Courseplay/courseplay/files/4315569/log.txt)

I have no idea what the difference is between what I did on this test and what I have been trying to do all day on my main save game and lets play map. Why does it work on one and not the other?

**Comment by vemendoza** [comment]
@cwattyeso ... My problem in #5031 was using an old savegame. In new savegame, no problem too. I don't know why... So I rollback CP to version 00386.

**Comment by Tensuko** [comment]
> 
> 
> Just to clarify. The right side of the Trailers are pointing in the direction of the field?

This doesn't matter anymore, Trailers can be filled from both sides :) Since V387.

**Comment by Merrlin** [comment]
@Tensuko ja weiß ich, wollt nur nen Bug im neuen SelfUnload Code ausschließen.
So two differences are only one harvester and new map. IDK if the map (and Version) is the same. So it could be the Map itselfes that causes the exception.

**Comment by cwattyeso** [comment]
My main savegame is on the Fenton Forext 4X Map, the test I did last night that worked was on the Lakeland Vale map. A Newer map but both from the same Map Author. I don't know why it works on one but not the other, went back to my original map and save just now tried to harvest a field with one harvest and one trailer. Tried doing it with Self Unload On and Off, Some of the other settings in Advanced On and Off. Both times I started getting LUA errors, but weirdly neither times when the Harvester was full. It just randomly as it was harvesting down a row, started erroring, the harvester kept driving at this point without harvesting, the in game camera wouldn't follow the harvester. I had to tab to it and stop the courseplay driver, at which point it started dumping straw everywhere. The second time it happened the Harvester had got to the end of a row, still wasn't full or needed to be emptied, but it decided to drive off the field and halfway across the map, before again I tabbed to it and stopped it. I only realised it had glitched again because the tractor I was using to bale straw at the same time on the same field, suddenly stopped picking up the straw despite everything being on. Quick pulled up the console and saw the LUA errors spamming again, so went to the combine and stopped it. So today on my original map/save game can not get harvesting to work, never mind self unloading.

On the new map and save I tested last night, the harvester used was the base game giants New Holland CR10.90 and the Tractor & Trailer was the John Deere 8R and the TAW30. Again base game equipment. Is the fact on my other save I'm trying to use DLC Vehicles and Equipment causing the problem, because when I check the log I see Courseplay really seems to dislike the ClaasPack.dlc stuff giving errors to to things from that like the Xerion 5000 and the Lexion 8900. Also because I have also tried using Modded Vehicles and Trailers with no joy, does Courseplay not like Modded Harvesters or Trailers at this point in time?

**Comment by cwattyeso** [comment]
Here are a couple of video clips from what I experienced earlier today, with the Courseplay Errors coming on the Harvester even before it had got full and would need to unload.

https://youtu.be/gadgjwf00Ds

https://youtu.be/WXxwVg6ftqI

Is there any debug modes I should try to enable whilst doing a course, which might generate some information in the log that might help better identify the cause of the problem and what might be the needed fix?

**Comment by Merrlin** [comment]
So it seems the error occurs when Combine hits the Fieldborder. Maybe there is a bug with that on Fenton Forrest. You can try to generate a course with Headlands as a workaround.
About the debugchannels @Tensuko knows better Wich one to use.

**Comment by cwattyeso** [comment]
@Merrlin The thing is Harvesting had been working fine on the Fenton Forest Map in all the time I have been playing and producing my series on it. I had used Courseplay without any issues previously across all the fields. I have even had times when I had Three Harvesters, working on the same fields in multiple tool/vehicle courses without problem.

I took quite a break from Farming Simulator 19 for a about a month and in that time my version of Courseplay went from 362 to 395. However I would not be able to say which version of Courseplay last worked for me on this save. As the problem with playing Seasons is, I'm not Harvesting all the time. So whilst I had been waiting for Harvest time on the Map, I probably had a few Courseplay updates from my last known working version.

I just remember in Version 362 being excited and looking forward to trying the Combine Self Unload Function which was fairly new in that, as an alternative to the the removal of the old Mode 7 from the FS17 version of Courseplay and the at time lack of of Mode 2 for Trailers.

Last Night with Version 395 I found a weird work around for my problem that let me harvest two fields, one with a single Claas Harvester and one with Two Tribines. What I did was have the normal Giants Workers drive the Harvesters and then used Courseplay and the new Mode 2 to operate the Trailers. The Trucks would drive to and unload the AI workers when they got full, and the AI workers would harvest pretty much 95% of the fields with only some missed bits on edges or corners, but without any LUA errors or breaking the game.

**Comment by vemendoza** [comment]
@cwattyeso did you tried cp 00386 or 00387 ..??  After this has a big change in code of CP for this functions

**Comment by cwattyeso** [comment]
No I didn't try 386 or 387. As I say when I first encountered the issue the other day I was using version 362, I came to the site here to check if there was an update for Courseplay and saw it was now on 395, so downloaded and installed that replacing my 362 version. I had not used or tried anything in between. The only thing I can confirm is I get the same problem in version 362 and 395, so would likely assume that if I downloaded and tried any version in between I would probably still have the same problem. My only thing to possibly check is to go back to version 344 I still have a copy of on this computer and try that and see if that was my last "Good" version where Harvester were able to run a harvest course without creating script errors.

**Comment by Tensuko** [comment]
It doesn't help if you try super old Versions.
Try V395, activate Channel 7 9 12 and 14 and turn on the 3 dots in CP HUD Settings when you make such nice Videos.

**Comment by cwattyeso** [comment]
Okay I'll give this a go later with the those Debug Channels activated.

**Comment by cwattyeso** [comment]
@Tensuko okay done a test this evening, with the debug channels you asked for. Here is the log file for that first of all. I deleted the log prior to starting the game, so the only information in it is relevant to this session and this test.
[log.txt](https://github.com/Courseplay/courseplay/files/4325855/log.txt)

I recorded and have uploaded a video to YouTube, it'll be at this link: https://youtu.be/SWbtD47_3Nc however it may require a little time for YouTube to process it into High Definition. I also apologise its a long video. 

I started the test with the Harvester set to try and self unload in a trailer parked next to the field. This caused script errors as soon as the harvester got full.
I stopped the Harvester, turned off self unload mode and let it carry on harvesting and this time went to the tractor and put that into Mode 2 to hopefully go and unload the harvester instead when it got full. However Mode 2 just caused all sorts of issues and quite frankly by the end of the video I was rage quitting. It just doesn't work no matter what I try doing. At this point it'd frankly be easier for me to harvest the fields manually.

**Comment by cwattyeso** [comment]
Just a quick question, does the latest version of Courseplay 362, 395, 397 etc that I have tried in the last couple of days not work with courses that were created and saved with much earlier versions Courseplay from the 200-300 range?

I've just noticed in game that if I load courses for example one where a Krone Mower mows a Grass Field, when I try to start the Courseplay Driver he just sits at the start of the course generating a similar script error to what I get with the Harvesters.

So is my problem something to do with my previously created and saved courses not being compatible with the new versions of courseplay?

**Comment by Tensuko** [comment]
I just watched your video.
I might have an Idea but I need to test that... Well wasn't the case (surrounded by fruit, wich is ofc not done IRL to start harvest in the middle section...

Your savegame already starts with a callstack at line 970, got another callstack at Line 1538 and the one when CP wants to go for unload starts in Line 6623 followed by the Line:
2020-03-12 16:42   dataS/scripts/network/Server.lua (58) Error: Running LUA method 'update'.
2020-03-12 16:42 D:/FS19 Mod Folder Manager/Old Family Farm/Courseplay_6_01_00395/course-generator/PathfinderUtil.lua:57: attempt to perform arithmetic on local 'rootToDirectionNodeDistance' (a nil value)

You can let the Combine harvest without unload and drive the unloader by yourself.
For the Xerion unloading part of your video... yea this have to be fixed and yes this is not easy to do, else we would be done with this Mode already.
I am not sure if you are informed about #3311 and #4286 but the current Mode2, the old one is not working correctly and is based on the super old system with 7 years old code...

**Comment by cwattyeso** [comment]
Yeah I know the Mode 2 was old code in Courseplay for FS19, I spent many hours telling people on Giants Forums and on Steam Forums about this when they would make posts complaining Courseplay wasn't working. I know that work on Mode 2 has recently only been added as Alpha. Before this I have tried many different ways of trying to work around this problem, but the removal of the Mode 7 from the FS17 Version of Courseplay made that difficult. So I was excited by the new Self Unload Options and the Alpha Mode 2, clearly the Mode 2 still needs a lot of work and is too frustrating to use with tractors driving constantly in circles or driving into the harvester.

As for the Callstack at 970 and 1538 they are being generated by the Xerion Tractor in the Claas DLC

2020-03-12 16:28 pdlc/claasPack/vehicles/claas/xerion5000/xerion5000.i3d (779.90 ms)
2020-03-12 16:28 LUA call stack:
2020-03-12 16:28   dataS/scripts/i3d/I3DUtil.lua (20) : printCallstack
2020-03-12 16:28   dataS/scripts/i3d/I3DUtil.lua (79) : checkChildIndex
2020-03-12 16:28   dataS/scripts/vehicles/specializations/AnimatedVehicle.lua (197) : indexToObject
2020-03-12 16:28   dataS/scripts/vehicles/specializations/FoliageBending.lua (215) : superFunc
2020-03-12 16:28   dataS/scripts/vehicles/specializations/AnimatedVehicle.lua (162) : loadAnimationPart
2020-03-12 16:28   dataS/scripts/vehicles/specializations/AnimatedVehicle.lua (93) : loadAnimation
2020-03-12 16:28   dataS/scripts/vehicles/SpecializationUtil.lua (30)
2020-03-12 16:28   dataS/scripts/vehicles/Vehicle.lua (615) : raiseEvent
2020-03-12 16:28   =D:/FS19 Mod Folder Manager/Old Family Farm/Courseplay_6_01_00395/register.lua (114) : superFunc
2020-03-12 16:28   dataS/scripts/i3d/I3DManager.lua (118) : asyncCallbackFunction
2020-03-12 16:28   Error: Index not found: 0>0|6|0|0|13|0
2020-03-12 16:28   Error: Failed to find child 13 from node cabinRotSupension, only 13 childs given

I've seen it sometimes pop up when driving and operating other vehicles form the DLC. I assume this is something Giants need to fix, or something Courseplay needs to fix to not conflict with the DLC and cause these errors. Is Courseplay updated to work with the updated Class DLC

I have the Steam version of the game, so automatically got the Claas DLC Update installed a couple of weeks ago, so unfortunately I can't remove it and go back to the pre-patch version of the DLC to see if this fixes these errors.

**Comment by cwattyeso** [comment]
@Tensuko I have posted a couple of times on the forums over at Giants regarding errors I'm seeing in the log related to the Claas DLC and how I'm experience issues in other games without Courseplay and even on servers I'm noticing some weird issues like for example in Multiplayer the Tracks on the Harvester don't rotate. However as is so often the case no one Officially from Giants every responds or provides answers.

**Comment by Tensuko** [comment]
So it is more a Claas DLC Issue ?
Well I will take a closer look to this but this would be real frustating if GIANTS isn't able to create bug free content when several Modders are able to do this....

Thanks for your Information!

**Comment by cwattyeso** [comment]
@Tensuko well it makes you wonder if it's a DLC Issue, when I'm experiencing errors in the game and log with vehicles from the Claas DLC, whether I'm using Courseplay or other mods, which makes me think their update has created way more problems then it meant to fix. Before the Claas Update, I never had any errors in my log referencing Claas Vehicles.

Now if I have a Claas Tractor and a Tool, load a previously saved and working course for a field, tell it to start at first waypoint, it doesn't move and just generates the LUA Error very similar to the one the Harvester has when it gets full and needs to unload. So I have been playing the last few days doing everything manual as I couldn't get any courses to run or work. Now I think perhaps I should have tried a different brand of vehicle and seen if they'd run the courses.

Also when I did a test the other evening on a different map, using the New Holland Harvester, and the John Deere 8R with the TAW30 Trailer, the Harvester worked fine the entire time, it self unloaded perfectly 3 times on one field. When I did another field, I left self unload off, and instead put the John Deere 8R into Mode 2 and it worked okay, better then in my video for definite with only one occasion where the Harvester turned left at the end of a field after finishing a row and crashed into the Truck which was driving next to it in the crop having just emptied it.

Another thing I now recall is on this save where I did the test, I didn't even have the Claas DLC selected or enabled when I started the game. So none of that stuff got loaded. Could be someone needs to interrogate Giants and find out what they've done and broken.
